include/viennals/lsCompareChamfer.hpp

```diff
@@ -0,0 +1,339 @@
+      return false;
+    }
+
+    return true;
+  }
+
+public:
+  CompareChamfer() {
+    static_assert(
+        D == 2,
+        "CompareChamfer is currently only implemented for 2D level sets.");
+  }
+
+  CompareChamfer(SmartPointer<Domain<T, D>> passedLevelSetTarget,
+                 SmartPointer<Domain<T, D>> passedLevelSetSample)
+      : levelSetTarget(passedLevelSetTarget),
+        levelSetSample(passedLevelSetSample) {
+    static_assert(
+        D == 2,
+        "CompareChamfer is currently only implemented for 2D level sets.");
+  }
+
+  /// Set the target level set
+  void setLevelSetTarget(SmartPointer<Domain<T, D>> passedLevelSet) {
+    levelSetTarget = passedLevelSet;
+  }
+
+  /// Set the sample level set
+  void setLevelSetSample(SmartPointer<Domain<T, D>> passedLevelSet) {
+    levelSetSample = passedLevelSet;
+  }
+
+  /// Set output mesh for target surface points with distance data
+  void setOutputMeshTarget(SmartPointer<Mesh<T>> passedMesh) {
+    outputMeshTarget = passedMesh;
+  }
+
+  /// Set output mesh for sample surface points with distance data
+  void setOutputMeshSample(SmartPointer<Mesh<T>> passedMesh) {
+    outputMeshSample = passedMesh;
+  }
+
+  /// Apply the Chamfer distance calculation
+  void apply() {
+    // Check compatibility
+    if (!checkCompatibility()) {
+      // Set NaN results to indicate failure
+      forwardDistance = std::numeric_limits<T>::quiet_NaN();
+      backwardDistance = std::numeric_limits<T>::quiet_NaN();
+      chamferDistance = std::numeric_limits<T>::quiet_NaN();
+      rmsChamferDistance = std::numeric_limits<T>::quiet_NaN();
+      maxDistance = std::numeric_limits<T>::quiet_NaN();
+      numTargetPoints = 0;
+      numSamplePoints = 0;
+      return;
+    }
+
+    // Ensure both level sets have sufficient width for surface extraction
+    constexpr int minimumWidth = 2;
+
+    auto workingTarget = levelSetTarget;
+    auto workingSample = levelSetSample;
+
+    if (levelSetTarget->getLevelSetWidth() < minimumWidth) {
+      workingTarget = SmartPointer<Domain<T, D>>::New(levelSetTarget);
+      Expand<T, D>(workingTarget, minimumWidth).apply();
+      Logger::getInstance()
+          .addInfo("CompareChamfer: Expanded target level set to width " +
+                   std::to_string(minimumWidth) + " for surface extraction.")
+          .print();
+    }
+
+    if (levelSetSample->getLevelSetWidth() < minimumWidth) {
+      workingSample = SmartPointer<Domain<T, D>>::New(levelSetSample);
+      Expand<T, D>(workingSample, minimumWidth).apply();
+      Logger::getInstance()
+          .addInfo("CompareChamfer: Expanded sample level set to width " +
+                   std::to_string(minimumWidth) + " for surface extraction.")
+          .print();
+    }
+
+    // Extract surface meshes
+    auto targetSurfaceMesh = SmartPointer<Mesh<T>>::New();
+    auto sampleSurfaceMesh = SmartPointer<Mesh<T>>::New();
+
+    ToSurfaceMesh<T, D>(workingTarget, targetSurfaceMesh).apply();
```

Comment:
I would use ToDiskMesh here since you only use surface nodes. It is way faster than ToSurfaceMesh